core/arch/x86_code_test.c

```diff
@@ -324,6 +324,73 @@ unit_test_get_zmm_caller_saved()
         0);
 }
 
+static void
+unit_test_get_opmask_caller_saved()
+{
+    /* While DynamoRIO's dr_opmask_t type is 8 bytes, the actual machine register is
+     * really only 8 bytes if the processor and OS support AVX512BW. Otherwise it is
+     * 2 Bytes.
+     */
+    dr_opmask_t ref_buffer[MCXT_NUM_OPMASK_SLOTS];
+    dr_opmask_t get_buffer[MCXT_NUM_OPMASK_SLOTS];
+    ASSERT(sizeof(dr_opmask_t) == OPMASK_REG_SIZE);
+    uint base = 0x0000348e;
+
+#        ifdef __AVX512BW__
+    /* i#1312: Modern AVX-512 machines support AVX512BW which extends the OpMask registers
+     * to 8 bytes. The right compile flags must then to be used to compile this test, and
+     * the type will be __mmask64. Also DynamoRIO's get_opmask_caller_saved has to
+     * dynamically switch dependent on a proc_ flag indicating AVX512BW is enabled.
+     */
+    FAIL();
```

Comment:
I think it might be a little nicer to have have a compile-time error for this. Can this be an #error, or is there a case I'm missing where this file must compile without __AVX512BW__? By my read, if you compile the test with __AVX512F__ but without __AVX512BW__ this test will always fail when run, but I could be missing something.

Reply:
Good idea. Changed FAIL() into #error.

> By my read, if you compile the test with AVX512F but without AVX512BW this test will always fail when run, but I could be missing something.

In future, the (also still missing-) proc_ test for AVX-512 (i.e. future ZMM_ENABLED()) will also test for AVX512BW support of processor + OS. If there is AVX512BW support, the x86_code_test module will be compiled with sth. like -march=skylake-avx512 instead of -mavx. The compiler provides the standard AVX512* flags like AVX512BW. At this point, the test needs to get fixed. Basically the #error is a reminder for this to happen.